concurrency/runner.go

```diff
@@ -60,47 +61,37 @@ func ForEachUser(ctx context.Context, userIDs []string, concurrency int, userFun
 	return errs.Err()
 }
 
-// ForEach runs the provided jobFunc for each job up to concurrency concurrent workers.
+// ForEachJob runs the provided jobFunc for each job index in [0, jobs) up to concurrency concurrent workers.
 // The execution breaks on first error encountered.
-func ForEach(ctx context.Context, jobs []interface{}, concurrency int, jobFunc func(ctx context.Context, job interface{}) error) error {
-	if len(jobs) == 0 {
+func ForEachJob(ctx context.Context, jobs int, concurrency int, jobFunc func(ctx context.Context, idx int) error) error {
+	if jobs == 0 {
 		return nil
 	}
 
-	// Push all jobs to a channel.
-	ch := make(chan interface{}, len(jobs))
-	for _, job := range jobs {
-		ch <- job
-	}
-	close(ch)
+	indexes := atomic.Int64{}
+	indexes.Add(int64(jobs))
 
 	// Start workers to process jobs.
 	g, ctx := errgroup.WithContext(ctx)
-	for ix := 0; ix < math.Min(concurrency, len(jobs)); ix++ {
+	for ix := 0; ix < math.Min(concurrency, jobs); ix++ {
 		g.Go(func() error {
-			for job := range ch {
+			for {
+				idx := int(indexes.Dec())
```

Comment:
This changes order of processed jobs. It shouldn't matter too much, but may be surprising to user (eg. when each job prints some output).

Reply:
Easy to fix: https://github.com/grafana/dskit/pull/113/commits/bae0d79bcd84ce6ab7a6f179c229a25fab4b8ed3